src/casts/specify-representation.rkt

```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
-  (sr-alloc "GVectT" TYPE-GVECT-TAG
-            `(("type" . ,(compile-type-static/dynamic ty)))))
-
-(: sr-type-mref (All (A) (A -> D0-Expr) A -> D0-Expr))
-(define (sr-type-mref compile-type-static/dynamic ty)
-  (sr-alloc "MRefT" TYPE-MREF-TAG
-            `(("type" . ,(compile-type-static/dynamic ty)))))
-
-(: sr-type-mvect (All (A) (A -> D0-Expr) A -> D0-Expr))
-(define (sr-type-mvect compile-type-static/dynamic ty)
-  (sr-alloc "MVectT" TYPE-MVECT-TAG
-            `(("type" . ,(compile-type-static/dynamic ty)))))
+(define-syntax (memory-layout$ stx)
+  (syntax-case stx ()
+    [(_ namespace name tag (field* cval* mode*) ...)
+     (begin
+       (define (id->string x) (symbol->string (syntax->datum x)))
+       (define namespace-string (id->string #'namespace))
+       (define namespace-string-caps (string-upcase namespace-string))
+       (define name-string (id->string #'name))
+       (define qualified-upcase-name
+         (format "~a-~a" namespace-string-caps (string-upcase name-string)))
+       (define (gen-index/offset field mode)
+         (define (gen-id param)
+           (format-id stx param qualified-upcase-name (string-upcase field)))
+         (if (eq? mode 'single) (gen-id "~a-~a-INDEX") (gen-id "~a-~a-OFFSET")))
+       (define (gen-data-id x) (format-id stx "data:~a" x))
+       (define field-string* (map syntax->datum (syntax->list #'(field* ...))))
+       (define mode-symbol* (map syntax->datum (syntax->list #'(mode* ...))))
+       (define index-def* (map gen-index/offset field-string* mode-symbol*))
+       (define index-data-def* (map gen-data-id index-def*))
+       (define-values (field-val* temp*)
+         (for/fold ([p* '()] [t* '()])
+                   ([c (syntax->list #'(cval* ...))])
+           (let ([cdtm (syntax->datum c)])
+             (if cdtm
+                 (values (cons (Quote cdtm) p*) t*)
+                 (with-syntax ([a (generate-temporary)])
+                   (values (cons #'a p*) (cons #'a t*)))))))
+       (define (all-but-last l)
+         (cond [(empty? l) (error 'all-but-last "empty list")]
+               [(empty? (rest l)) empty]
+               [else (cons (first l) (all-but-last (rest l)))]))
+       (with-syntax* ([func-alloc-name
+                       (format-id
+                        stx "sr-~a-~a" namespace-string name-string)]
+                      [func-huh-name
+                       (format-id stx "~a-~a?" namespace-string name-string)]
+                      [namespace-mask-def
+                       (format-id stx "~a-TAG-MASK" namespace-string-caps)]
+                      [tag-def (format-id stx "~a-TAG" qualified-upcase-name)]
+                      [sname-string name-string]
+                      [(sfield-string* ...) field-string*]
+                      [(sindex-def* ...) index-def*]
+                      [(sindex-data-def* ...) index-data-def*]
+                      [tag-data-def (gen-data-id #'tag-def)]
+                      [(sindex-val* ...)
```

Comment:
This is the same as `(datum->syntax stx (range ...))`

Reply:
Done.